site-src/geps/gep-1323.md

```diff
@@ -0,0 +1,274 @@
+    // +kubebuilder:validation:MaxItems=16
+    Set []HTTPHeader `json:"set,omitempty"`
+
+    // Add adds the given header(s) (name, value) to the request
+    // before the action. It appends to any existing values associated
+    // with the header name.
+
+    // +optional
+    // +listType=map
+    // +listMapKey=name
+    // +kubebuilder:validation:MaxItems=16
+    Add []HTTPHeader `json:"add,omitempty"`
+
+    // Remove the given header(s) from the HTTP request before the action. The
+    // value of Remove is a list of HTTP header names. Note that the header
+    // names are case-insensitive (see
+    // https://datatracker.ietf.org/doc/html/rfc2616#section-4.2).
+    // +optional
+    // +kubebuilder:validation:MaxItems=16
+    Remove []string `json:"remove,omitempty"`
+}
+```
+
+Given the fact that this functionality is offered by only a few projects that currently implement Gateway API when using their own traffic routing CRDs, it’s better to support `ResponseHeaderModifier` as an _Extended_ feature, unlike `RequestHeaderModifier` which is a _Core_ feature. This will also not increase the difficulty of implementing Gateway API for any future ingress or service mesh.
+
+## Usage
+Adding support for this unlocks a lot of real world use cases. Let’s review a couple of them:
+
+* A team has a frontend web app, along with two different versions of their backends exposed as Kubernetes services. If, the frontend needs to know which backend it’s talking to, this can be easily achieved without any modifications to the application code.
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: HTTPRoute
+metadata:
+  name: http-response-header
+spec:
+  hostnames:
+    - response.header.example
+  rules:
+    - backendRefs:
+      - name: example-svc-beta
+        weight: 50
+        port: 80
+        # set a custom header for all responses being sent from the beta build of the backend server.
+        filters:
+           - type: ResponseHeaderModifier
+             responseHeaderModifier:
+              add:
+                name: build
+                value: beta
+      - name: example-svc-stable
+        weight: 50
+        port: 80
+```
+
+* Cookies can be automatically injected into the response of services. This can enable services to identify users that were redirected to a certain backend.
+
+```yaml
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: HTTPRoute
+metadata:
+  name: http-response-header
+spec:
+  hostnames:
+    - response.header.example
+  rules:
+    # match against any requests that has the cookie set due to the below rule
+    - matches:
+      - headers:
+        type: Exact
+        name: Cookie
+        value: user=insider
+      backendRefs:
+        - name: foo-svc
+        port: 8080
+
+    - filters:
+      - type: ResponseHeaderModifier
+        # set cookies for all requests being forwarded to this service
+        responseHeaderModifier:
+          set:
+            name: Set-Cookie
```

Comment:
According to https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie, "To send multiple cookies, multiple Set-Cookie headers should be sent in the same response."   Will it be possible to set multiple headers with the same name?

Reply:
You can use `add` instead of `set`: https://github.com/kubernetes-sigs/gateway-api/blob/30f2983e9f01368adff82772c4d9178fcdb26344/apis/v1beta1/httproute_types.go#L730

Reply:
I think the validation will prevent it though - it will reject any attempt to set or add two headers with the same name
```
    // +listType=map
    // +listMapKey=name
  ```
at the same time, it is probably will be very uncommon requirement

Reply:
i'm not 100% sure, but i think that'll happen when you have the same header name (the key) present in both `set` _and_ `add`. This conclusion of mine stems from https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules
> map: X + Y performs a merge where the array positions of all keys in X are preserved but the values are overwritten by values in Y when the key sets of X and Y intersect. Elements in Y with non-intersecting keys are appended, retaining their partial order.